### PR TITLE
New HMAC SHA256 auth method for MapTiler 

### PR DESCRIPTION
**Pull request by vcloarec**
A new auth method specific for MapTiler with a token encrypted  with HMAC-SHA256

image

Funded by [MapTiler](https://www.maptiler.com/)

### CODE REVIEW

File: src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    QgsDebugMsg( QStringLiteral( "Updating old style auth method config" ) );
+  }
+
+  // NOTE: add updates as method version() increases due to config storage changes
+}
+
+QgsAuthMethodConfig QgsAuthMapTilerHmacSha256Method::getMethodConfig( const QString &authcfg, bool fullconfig )
+{
+  const QMutexLocker locker( &mMutex );
+  QgsAuthMethodConfig mconfig;
```
Comment (nyalldawson) [inline src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp:121]:
```suggestion
  QgsAuthMethodConfig config;
```

File: src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+  // check if it is cached
+  if ( sAuthConfigCache.contains( authcfg ) )
+  {
+    mconfig = sAuthConfigCache.value( authcfg );
+    QgsDebugMsg( QStringLiteral( "Retrieved config for authcfg: %1" ).arg( authcfg ) );
```
Comment (nyalldawson) [inline src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp:127]:
```suggestion
    QgsDebugMsgLevel( QStringLiteral( "Retrieved config for authcfg: %1" ).arg( authcfg ), 2 );
```

File: src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+  // cache bundle
+  putMethodConfig( authcfg, mconfig );
+
+  return mconfig;
+}
+
+void QgsAuthMapTilerHmacSha256Method::putMethodConfig( const QString &authcfg, const QgsAuthMethodConfig &mconfig )
+{
+  const QMutexLocker locker( &mMutex );
+  QgsDebugMsg( QStringLiteral( "Putting token config for authcfg: %1" ).arg( authcfg ) );
```
Comment (nyalldawson) [inline src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp:147]:
```suggestion
  QgsDebugMsgLevel( QStringLiteral( "Putting token config for authcfg: %1" ).arg( authcfg ), 2);
```

Also elsewhere, for all non-error debug messages

**Comment by nyalldawson** [comment]
You'll also need to update the various packaging scripts to include the new provider

File: src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp (hunk 3 above)
Comment (vcloarec) [reply src/auth/maptiler_hmacsha256/core/qgsauthmaptilerhmacsha256method.cpp]: done, also done in esri token auth where this code come from.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thanks, all looks good to me!

**Comment by wonder-sk** [comment]
To add some more context to this PR: normally providers of online raster/vector tiles use a "key" query parameter in URL to authenticate the user, e.g. `http://example.com/{z}/{x}/{y}.png?key=abcdefgh`. The problem with that is that such key is not really secret and it can be easily leaked and abused by someone else. A solution is to use "signed URLs". Provider creates a token that is composed of "public" and "secret" part, and each request gets signed like this:
1. take the original URL (e.g. `http://example.com/0/0/0.png`
2. add the public part of the token to the url: `http://example.com/0/0/0.png?key=xxxxxxx`
3. compute signature by mixing the URL with secret part of token + add `&signature=ijklmnopqrst` to the URL.

This way the signature is different for each request, and the server can verify that the signature is correct. And secret part of token is not revealed anywhere.

As a bonus, the token also does not get saved to QGIS project file because it is stored in the auth manager.

**Pull request merged**